[FIX] Added plain text pasting and fixed issues with image pasting.

charlwillia6 opened the pull request:
## Description ✏️

Closes [#144](https://github.com/trypear/pearai-app/issues/114)

Fixed issues with pasting links and meta data getting pasted instead of the raw link.
There was also issues with pasting screenshots as images.  This fixes those issues.
Fixed this issue:
![image](https://github.com/user-attachments/assets/beeb7ffd-1ae4-4999-8fbd-1c887901514a)

Moved logging `[2024-11-15T16:40:45] [LOG] Pasting image` so that this log only prints when actually pasting an image.

@Himanshu-Singh-Chauhan will need to test and verify.

## Checklist ✅

- [x] I have added screenshots (if UI changes are present).
- [x] I have done a self-review of my code.
- [x] I have manually tested my code (if applicable).

nang-dev commented:
<img width="630" alt="Screenshot 2024-08-26 at 4 03 44 PM" src="https://github.com/user-attachments/assets/36756beb-8017-4a29-a714-2e20ee0fc65a">
<img width="563" alt="Screenshot 2024-08-26 at 3 30 36 PM" src="https://github.com/user-attachments/assets/5dd54951-eb0d-4ac2-ae37-6a84b3f023ef">

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Fix: improve image and text paste handling in TipTapEditor

- Add check for image content before processing paste events
- Prevent default paste behavior only when handling images
- Insert images at cursor position instead of document start
- Add plain text paste handler to strip formatting
- Clean up image paste logic with better type checking and error handling

charlwillia6 committed:
Updated message

charlwillia6 commented:
@Himanshu-Singh-Chauhan Please test

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main' into charlwillia6/fix-issues-with-pasting

charlwillia6 committed:
Fixed image positioning and a bug where newlines were being removed when a user starts typing after pasting.

Himanshu-Singh-Chauhan committed:
delete selected text

Himanshu-Singh-Chauhan commented:
tested, works. awesome work. 🔥

Fryingpannn commented on gui/src/components/mainInput/TipTapEditor.tsx:
```diff
@@ -327,31 +327,44 @@ const TipTapEditor = memo(function TipTapEditor({
+
+                  // Only log and process if we actually have an image
+                  if (hasImageItem) {
+                    console.log("Pasting image");
```
console log!

charlwillia6 replied:
This was already in the code. I only moved it to inside this logic. @Fryingpannn do you want me to delete it?

Fryingpannn committed:
Remove console log

Fryingpannn commented:
https://github.com/user-attachments/assets/0c6e94ab-38c9-475c-bac2-58445ce69c47

copy pasting over existing text not working

Himanshu-Singh-Chauhan committed:
Merge branch 'charlwillia6/fix-issues-with-pasting' of https://github.com/charlwillia6/pearai-submodule into charlwillia6/fix-issues-with-pasting

Himanshu-Singh-Chauhan commented:
copy pasting over existing text not working should be fixed now

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main' into charlwillia6/fix-issues-with-pasting

charlwillia6 commented:
> copy pasting over existing text not working should be fixed now

Works for me!  Excellent job @Himanshu-Singh-Chauhan!

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main' into charlwillia6/fix-issues-with-pasting

charlwillia6 commented:
> Screen.Recording.2024-11-17.at.5.03.33.PM.mov 
> copy pasting over existing text not working

This has been resolved by @Himanshu-Singh-Chauhan

Fryingpannn commented:
I can't reproduce the error in the description ): are u able to repro it? if so what are the steps?

charlwillia6 commented:
> I can't reproduce the error in the description ): are u able to repro it? if so what are the steps?

I don't think it is platform dependent. And it is an error that doesn't happen all the time. The change I made is for a more aggressive check for what the user is trying to paste. Before, the code was not properly checking whether what was being pasted was an image or text. Sometimes it would think it was an image when it was really text, and sometimes it didn't register screenshots properly as images, so it wouldn't even add them to the input box. If I imported them as a file from path, it would work fine. 

I think two things are happening. Sometimes, when you add an image to the chat input box, after that it would think everything was an image that you were trying to paste. So, it would fail until you reloaded the webviews or sometimes it would reset if you created a new session. Second, it didn't always properly know if what you were pasting was an image or text. 

My changes just do a better job if checking what is being pasted, and makes sure that doesn't treat it as an image if it isn't one.

Fryingpannn commented:
Hmm I don't think I can merge this until I can reproduce an error we currently have and see this fixes it. I'm not sure if fully necessary. Cause rn it works fine for me? Also now with this change we're gonna have 2 paste actions (`paste` and `handlePaste`), seems not clean to me.

Most important thing rn for this is to understand how to repro the error in the screenshot u sent. Then tackle that issue specifically.

charlwillia6 commented:
> Hmm I don't think I can merge this until I can reproduce an error we currently have and see this fixes it. I'm not sure if fully necessary. Cause rn it works fine for me? Also now with this change we're gonna have 2 paste actions (`paste` and `handlePaste`), seems not clean to me.
> 
> Most important thing rn for this is to understand how to repro the error in the screenshot u sent. Then tackle that issue specifically.

I will try to find a way to reproduce. This is affecting a lot of users, so it is a necessary fix. Maybe it's not an issue on Mac. I have no idea.  I will see if there is any cleanup I can do too. 

Remember, we are also fixing two separate issues with this PR. Image pasting and plain text pasting.

charlwillia6 commented:
> Hmm I don't think I can merge this until I can reproduce an error we currently have and see this fixes it. I'm not sure if fully necessary. Cause rn it works fine for me? Also now with this change we're gonna have 2 paste actions (`paste` and `handlePaste`), seems not clean to me.
> 
> Most important thing rn for this is to understand how to repro the error in the screenshot u sent. Then tackle that issue specifically.

So, the reason that there are two different "paste" handlers, is because one really just handles images, and the other handles general pasting, specifically for this PR, plain text pasting. They both do two different things.  While we could combine them, that would not really be ideal.  I also asked Claude:

Looking at both paste handlers, they serve different but complementary purposes:

1. The first handler (lines 329-366) is specifically for handling image paste events. It's part of the Image extension configuration and only handles cases where an image is being pasted.

2. The second handler (lines 505-532) is a general paste handler in editorProps that handles text pasting, with special handling for maintaining line breaks using hardBreak nodes.

They could be combined, but keeping them separate actually makes sense because:

1. Separation of Concerns: The Image extension handler is specifically focused on image-related functionality, while the general handler deals with text formatting.

2. Extension Architecture: The first handler is part of TipTap's extension system, while the second is part of the core editor properties.

3. Control Flow: The current setup allows for clean handling - if an image is being pasted, the Image extension handler takes care of it. If it's text, the general handler handles it.  Also, you can reproduce this error easily with Haiku buy adding an image to the chat input box.  But, that is probably because Haiku doesn't support images?  I am not sure.

That said, they could be combined into a single handler in editorProps like this:

```typescript
handlePaste(view, event) {
  const items = event.clipboardData.items;
  const hasImageItem = Array.from(items).some(item => item.type.startsWith('image/'));
  
  if (hasImageItem) {
    for (const item of items) {
      if (!item.type.startsWith('image/')) continue;
      
      const file = item.getAsFile();
      if (!file) continue;

      if (modelSupportsImages(
        defaultModel.provider,
        defaultModel.model,
        defaultModel.title,
        defaultModel.capabilities,
      )) {
        event.preventDefault();
        
        handleImageFile(file).then((resp) => {
          if (!resp) return;
          
          const [img, dataUrl] = resp;
          const { schema } = view.state;
          const node = schema.nodes.image.create({
            src: dataUrl,
          });
          const tr = view.state.tr.insert(0, node);
          view.dispatch(tr);
        });
        return true;
      }
    }
  }

  // Handle text paste
  event.preventDefault();
  const text = event.clipboardData.getData('text/plain');
  const lines = text.split(/\r?\n/);
  const { tr } = view.state;
  const { schema } = view.state;
  let pos = view.state.selection.from;

  tr.delete(view.state.selection.from, view.state.selection.to);

  lines.forEach((line, index) => {
    if (index > 0) {
      tr.insert(pos++, schema.nodes.hardBreak.create());
    }
    tr.insertText(line, pos);
    pos += line.length;
  });

  view.dispatch(tr);
  return true;
}
```

However, keeping them separate is likely better for code organization and maintainability. The current structure makes it clearer which code handles which functionality and follows TipTap's extension-based architecture.

@Fryingpannn The one paste handle is extending the image handling of the editor.  This is more for when you add an image using the file import, but also for pasting.  It is also doing it by manipulating DOM events. The handlePaste function is for general pasting.  Like for text.  They do two separate things.  There is just a check in the handlePaste function to make sure it isn't an image.  It doesn't make sense to combine them.

Also, the image pasting part of this doesn't just fix the issue above in the screenshot (bad request), it also fixes the issue with pasting screenshots with CTRL+V.  That doesn't work most of the time.  If you need me to send you a video of that not working, then I will.  Let me know if you can reproduce that issue.

Let me know what I can do to get this approved.  Thanks.

charlwillia6 committed:
Found a bug when pasting images with CTRL+V

Fryingpannn commented:
Will get to your comment soon.

Adding another comment about something I remembered when testing earlier: when I copy pasted image from google docs it didn't work. Which works for me on my current version.

charlwillia6 commented:
> Will get to your comment soon.
> 
> Adding another comment about something I remembered when testing earlier: when I copy pasted image from google docs it didn't work. Which works for me on my current version.

Hmm. Can you do a git pull and try again? I found an issue and fixed.  There was a useEffect that was conflicting with the changes to the paste handlers.

Also, have @Himanshu-Singh-Chauhan test too with Windows please. Because on all versions that I have had of PearAI, copy and pasting almost never works for images. It is indeed an issue.

charlwillia6 commented:
Here is a recording showing that I took a screenshot and it is in my clipboard.  I try to use CTRL+V and it doesn't work.  CTRL+V for screenshots does not work 99% of the time.

[recording-2024-11-24-05-50-11.webm](https://github.com/user-attachments/assets/52f99995-d24e-47df-8836-24766094cd38)

charlwillia6 committed:
Updated useEffect so that the paste handlers handle paste operations

charlwillia6 commented:
Here is a screen recording of it working with this PR:



https://github.com/user-attachments/assets/544ca1af-544a-4964-9798-e8af973e6947

charlwillia6 committed:
Removed extra spaces.

charlwillia6 committed:
Combined logic into one paste handler.
Better checks for the selected model.

charlwillia6 commented:
@Fryingpannn I took your advice and created one paste handler so there was nothing conflicting.  Also, I made sure there was better checking for the selected model.  I think that one of the issues was that when the extension first loaded, the selected model wasn't known to the image paste handler, so pasting an image would fail due to a `return false`.  I think that this PR is in a good spot and has better checking for text and images.

charlwillia6 committed:
Removed spaces

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main' into charlwillia6/fix-issues-with-pasting

charlwillia6 commented:
@Himanshu-Singh-Chauhan - This PR can be closed.  It has been replaced by https://github.com/trypear/pearai-submodule/pull/259
